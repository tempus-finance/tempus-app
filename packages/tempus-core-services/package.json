{
  "name": "tempus-core-services",
  "version": "0.0.1",
  "private": true,
  "description": "Tempus core Services",
  "keywords": [
    "typescript",
    "boilerplate",
    "esbuild"
  ],
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "lint": "eslint src/ --ext .js,.jsx,.ts,.tsx",
    "test": "jest",
    "clean": "rm -rf dist build package",
    "ts-node": "ts-node",
    "docs": "typedoc --entryPoints src/index.ts",
    "build": "npm run clean && tsc -p tsconfig.json",
    "build-all": "npm run clean && npm run build && npm run esbuild-node && npm run esbuild-node",
    "esbuild-node": "esbuild src/index.ts --bundle --platform=node --minify --sourcemap=external --outfile=dist/esbuild/index.js",
    "esbuild-node:dev": "esbuild src/index.ts --bundle --platform=node --sourcemap=external --outfile=dist/esbuild/index.js",
    "esbuild-node:watch": "esbuild src/index.ts --bundle --platform=node --watch --sourcemap=external --outfile=dist/esbuild/index.js"
  },
  "dependencies": {
    "@ethersproject/providers": "^5.5.2",
    "@types/jest": "^24.9.1",
    "@types/node": "^12.20.15",
    "axios": "^0.22.0",
    "date-fns": "^2.28.0",
    "ethers": "^5.4.6"
  },
  "devDependencies": {
    "esbuild": "^0.11.11",
    "eslint": "^7.22.0",
    "jest": "^26.6.3",
    "ts-jest": "^26.5.4",
    "ts-node": "^9.1.1",
    "typedoc": "^0.20.35"
  },
  "peerDependencies": {
    "@typescript-eslint/eslint-plugin": "^4.28.1",
    "@typescript-eslint/parser": "^4.28.1",
    "eslint-config-airbnb-typescript": "^12.3.1"
  }
}
