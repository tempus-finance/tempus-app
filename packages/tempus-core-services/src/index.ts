import ERC20ABI from './abi/ERC20.json';
import VaultABI from './abi/Vault.json';

export { ERC20ABI, VaultABI };

export type { ERC20 } from './abi';

export type {
  AvailableToDeposit,
  Chain,
  ChainConfig,
  Config,
  DynamicTempusPool,
  PoolShares,
  ProtocolDisplayName,
  ProtocolName,
  Ticker,
  NativeTokenTicker,
  TempusPool,
  TokenPrecision,
  TokenTypePrecision,
  YearnData,
  YearnDataApy,
  YearnDataToken,
} from './interfaces';

export type {
  DepositedEvent,
  ERC20TokenService,
  PoolBalanceChangedEvent,
  PoolBalanceChangedEventListener,
  RedeemedEvent,
  TempusAMMJoinKind,
  TempusAMMExitKind,
  TempusAMMService,
  TempusControllerService,
  TempusPoolService,
  TransferEventListener,
  VaultService,
  StatisticsService,
  StorageService,
  SwapEvent,
} from './services';

export type { Numberish } from './datastructures';

export { Decimal, ZERO, ONE, DEFAULT_DECIMAL_PRECISION } from './datastructures';

export { SwapKind } from './services';

export {
  getChainlinkFeed,
  getCoingeckoRate,
  getDefaultProvider,
  getERC20TokenService,
  initServices,
  getServices,
  getDefinedServices,
} from './services';

export {
  NumberUtils,
  DecimalUtils,
  capitalize,
  chainIdToChainName,
  chainIdHexToChainName,
  chainNameToHexChainId,
  chainToTicker,
  decreasePrecision,
  div18f,
  getCookie,
  getLibrary,
  getPastDaysNumber,
  getProvider,
  getProviderFromSignerOrProvider,
  getRangeFrom,
  getTokenPrecision,
  increasePrecision,
  isZeroString,
  mul18f,
  prettifyChainName,
  prettifyChainNameLong,
  rayToDai,
  reverseString,
  shortenAccount,
  wadToDai,
  wait,
  prettifyProtocolName,
} from './utils';

export * from './constants';
