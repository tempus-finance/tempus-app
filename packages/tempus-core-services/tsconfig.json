{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "declaration": true,

    "importHelpers": true,
    "experimentalDecorators": true,
    "sourceMap": true,
    "outDir": "./dist/",
    "types": ["node", "jest"]
  },
  "include": ["src/**/*.ts"],
  "exclude": ["node_modules"]
}
