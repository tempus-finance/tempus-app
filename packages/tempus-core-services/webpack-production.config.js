const path = require('path');
const webpack = require('webpack');

require('dotenv').config({ path: './.env.production' });

module.exports = {
  entry: './src/index.ts',
  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: 'ts-loader',
        exclude: [/node_modules/],
      },
    ],
  },
  resolve: {
    extensions: ['.ts', '.js'],
  },
  output: {
    path: path.resolve(__dirname, 'dist'),
    publicPath: '',
    filename: 'index.js',
    library: 'tempus-core-services',
    libraryTarget: 'umd',
  },
  mode: 'production',
  devtool: 'eval-source-map',
  externals: ['ethers', 'rxjs', 'date-fns', '@ethersproject/providers'],
  plugins: [
    new webpack.DefinePlugin({
      'process.env': JSON.stringify(process.env),
    }),
  ],
};
