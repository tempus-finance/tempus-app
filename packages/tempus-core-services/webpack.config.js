// Webpack uses this to work with directories
const path = require('path');
const webpack = require('webpack');

require('dotenv').config({ path: '../tempus-client_v3/.env.development' });

// This is the main configuration object.
// Here, you write different options and tell Webpack what to do
module.exports = {
  // Path to your entry point. From this file Webpack will begin its work
  entry: './src/index.ts',

  module: {
    rules: [
      {
        test: /\.tsx?$/,
        use: 'ts-loader',
        exclude: [/node_modules/],
        include: [path.resolve(__dirname, 'src')],
      },
    ],
  },
  resolve: {
    extensions: ['.ts', '.js'],
  },

  // Path and filename of your result bundle.
  // Webpack will bundle all JavaScript into this file
  output: {
    path: path.resolve(__dirname, 'dist'),
    publicPath: '',
    filename: 'index.js',
    library: 'tempus-core-services',
    libraryTarget: 'umd',
  },

  // Default mode for Webpack is production.
  // Depending on mode Webpack will apply different things
  // on the final bundle. For now, we don't need production's JavaScript
  // minifying and other things, so let's set mode to development
  mode: 'development',

  devtool: 'eval-source-map',

  watch: true,

  plugins: [
    new webpack.DefinePlugin({
      'process.env': JSON.stringify(process.env),
    }),
  ],
};
