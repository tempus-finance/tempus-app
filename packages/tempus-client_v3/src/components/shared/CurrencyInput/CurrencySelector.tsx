import { FC, useCallback, useState } from 'react';
import { Ticker } from 'tempus-core-services';
import Icon from '../Icon';
import CurrencySelectorItem from './CurrencySelectorItem';

export interface CurrencySelectorProps {
  currencies: Ticker | Ticker[];
  defaultCurrency?: Ticker;
  disabled?: boolean;
  onChange?: (selectedCurrency: Ticker) => void;
}

const CurrencySelector: FC<CurrencySelectorProps> = props => {
  const { currencies, defaultCurrency, disabled, onChange } = props;
  const hasMultipleCurrencies = currencies.constructor === Array && currencies.length > 1;

  const [selectedCurrency, setSelectedCurrency] = useState(defaultCurrency ?? (currencies[0] as Ticker));
  const [dropdownOpened, setDropdownOpened] = useState(false);

  const handleSelectorOpen = useCallback(() => setDropdownOpened(true), []);

  const handleSelectorClose = useCallback(() => setDropdownOpened(false), []);

  const handleSelectorItemClick = useCallback(
    (currency: Ticker) => {
      if (selectedCurrency !== currency) {
        onChange?.(currency);
      }

      setSelectedCurrency(currency);
      setDropdownOpened(false);
    },
    [onChange, selectedCurrency],
  );

  return (
    <div
      className={`tc__currency-input__currency-selector ${
        hasMultipleCurrencies ? 'tc__currency-input__currency-selector-multiple-choice' : ''
      } ${disabled ? 'tc__currency-input__currency-selector-disabled' : ''}`}
    >
      <CurrencySelectorItem
        currency={selectedCurrency}
        disabled={disabled}
        icon={hasMultipleCurrencies && <Icon variant="down-chevron" size="small" />}
        onClick={handleSelectorOpen}
      />
      {hasMultipleCurrencies && dropdownOpened && (
        <>
          <div className="tc__currency-input__currency-selector-backdrop" onClick={handleSelectorClose} />
          <div className="tc__currency-input__currency-selector-dropdown">
            {currencies.map((currency, index) => (
              <CurrencySelectorItem
                currency={currency}
                disabled={disabled}
                icon={index === 0 && <Icon variant="up-chevron" size="small" />}
                onClick={handleSelectorItemClick}
                key={currency}
              />
            ))}
          </div>
        </>
      )}
    </div>
  );
};

export default CurrencySelector;
