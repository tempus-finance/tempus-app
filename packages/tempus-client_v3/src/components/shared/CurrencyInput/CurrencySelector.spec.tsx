import { fireEvent, getAllByRole, render } from '@testing-library/react';
import { Ticker } from 'tempus-core-services';
import CurrencySelector, { CurrencySelectorProps } from './CurrencySelector';

const defaultProps = {
  currencies: ['ETH', 'stETH'],
};

const mockOnChange = jest.fn<void, [Ticker]>();

const subject = (props: CurrencySelectorProps) => render(<CurrencySelector {...props} />);

describe('CurrencySelector', () => {
  it('renders a currency selector', () => {
    const { getByRole, queryByText } = subject(defaultProps);

    const actualButton = getByRole('button');
    const actualText = queryByText(defaultProps.currencies[0]);

    expect(actualButton).not.toBeNull();
    expect(actualText).not.toBeNull();

    expect(actualButton).toMatchSnapshot();
    expect(actualText).toMatchSnapshot();
  });

  it('renders a disabled currency selector', () => {
    const { getByRole } = subject({ ...defaultProps, disabled: true });

    const actualButton = getByRole('button');

    expect(actualButton).not.toBeNull();
    expect(actualButton).toBeDisabled();

    expect(actualButton).toMatchSnapshot();
  });

  it('opens a currency selector with multiple currencies', () => {
    const { getAllByRole, getByRole, queryAllByRole, queryAllByText, queryByText } = subject({
      ...defaultProps,
      onChange: mockOnChange,
    });

    let actualEthButtons = getAllByRole('button', { name: defaultProps.currencies[0] });

    expect(actualEthButtons).not.toBeNull();
    expect(actualEthButtons).toHaveLength(1);

    fireEvent.click(actualEthButtons[0]);

    actualEthButtons = getAllByRole('button', { name: defaultProps.currencies[0] });
    const actualEthTexts = queryAllByText(defaultProps.currencies[0]);
    const availableStEthButton = getByRole('button', { name: defaultProps.currencies[1] });
    const availableStETHText = queryByText(defaultProps.currencies[1]);

    expect(actualEthButtons).not.toBeNull();
    expect(actualEthButtons).toHaveLength(2);

    expect(actualEthTexts).not.toBeNull();
    expect(actualEthTexts).toHaveLength(2);

    expect(availableStEthButton).not.toBeNull();

    expect(availableStETHText).not.toBeNull();

    // Change to the second currency
    fireEvent.click(availableStEthButton);

    expect(mockOnChange).toBeCalledTimes(1);
    expect(mockOnChange).toBeCalledWith(defaultProps.currencies[1]);

    const unavailableEthButtons = queryAllByRole('button', { name: defaultProps.currencies[0] });
    const unavailableEthText = queryByText(defaultProps.currencies[0]);
    const actualStEthButton = getByRole('button', { name: defaultProps.currencies[1] });
    const actualStEthText = queryByText(defaultProps.currencies[1]);

    expect(unavailableEthButtons).toHaveLength(0);
    expect(unavailableEthText).toBeNull();

    expect(actualStEthButton).not.toBeNull();
    expect(actualStEthText).not.toBeNull();
  });

  it('does not call `onChange` when the same currency is selected', () => {
    const { container, getByRole } = subject({
      ...defaultProps,
      onChange: mockOnChange,
    });

    const actualSelectorButton = getByRole('button');

    expect(actualSelectorButton).not.toBeNull();

    fireEvent.click(actualSelectorButton);

    const dropdownEthButton = container.querySelector(
      '.tc__currency-input__currency-selector-dropdown button:first-of-type',
    );

    expect(dropdownEthButton).not.toBeNull();

    fireEvent.click(dropdownEthButton);

    expect(mockOnChange).not.toBeCalled();
  });
});
