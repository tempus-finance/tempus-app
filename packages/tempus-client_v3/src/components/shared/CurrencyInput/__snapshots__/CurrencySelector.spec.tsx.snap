// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySelector renders a currency selector 1`] = `
<button
  class="tc__btn tc__currency-input__currency-selector-item "
  data-selected="false"
  type="button"
>
  <span
    class="tc__currency-input__currency-selector-item-content"
  >
    <svg
      class="tc__logo tc__logo-token-ETH"
      fill="none"
      height="40"
      viewBox="0 0 40 40"
      width="40"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20 40C31.0457 40 40 31.0457 40 20C40 8.9543 31.0457 0 20 0C8.9543 0 0 8.9543 0 20C0 31.0457 8.9543 40 20 40Z"
        fill="#627EEA"
      />
      <path
        clip-rule="evenodd"
        d="M11.25 20.275L20.6225 5V16.0875L11.25 20.275ZM20.6225 27.4588V34.9938L11.25 22.0201L20.6225 27.4588Z"
        fill="white"
        fill-rule="evenodd"
      />
      <path
        d="M20.625 25.7163L29.9962 20.275L20.625 16.09V25.7163Z"
        fill="white"
        fill-opacity="0.2"
      />
      <path
        clip-rule="evenodd"
        d="M20.625 5V16.0875L29.9962 20.275L20.625 5ZM20.625 27.4573V34.991L30.0025 22.0173L20.625 27.4573ZM20.6225 25.7163L11.25 20.275L20.6225 16.09V25.7163Z"
        fill="white"
        fill-opacity="0.6"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="tc__currency-input__currency-selector-item-label"
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      ETH
    </div>
  </span>
  <svg
    class="tc__icon tc__icon-down-chevron"
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      clip-path="url(#clip0_909_5439)"
    >
      <path
        d="M15.7387 5.36992L8.63082 11.7636C8.54823 11.8385 8.44996 11.898 8.34169 11.9385C8.23342 11.9791 8.11729 12 8 12C7.88271 12 7.76658 11.9791 7.65831 11.9385C7.55004 11.898 7.45177 11.8385 7.36918 11.7636L0.261296 5.36992C0.0939908 5.21942 -2.49304e-09 5.01531 0 4.80248C2.49304e-09 4.58965 0.0939908 4.38554 0.261296 4.23504C0.428601 4.08455 0.655515 4 0.89212 4C1.12873 4 1.35564 4.08455 1.52294 4.23504L8 10.0693L14.4771 4.23504C14.6444 4.08455 14.8713 4 15.1079 4C15.3445 4 15.5714 4.08455 15.7387 4.23504C15.906 4.38554 16 4.58965 16 4.80248C16 5.01531 15.906 5.21942 15.7387 5.36992Z"
        fill="#062330"
      />
    </g>
    <defs>
      <clippath
        id="clip0_909_5439"
      >
        <rect
          fill="white"
          height="16"
          width="16"
        />
      </clippath>
    </defs>
  </svg>
</button>
`;

exports[`CurrencySelector renders a currency selector 2`] = `
<div
  class="tc__currency-input__currency-selector-item-label"
  style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
>
  ETH
</div>
`;

exports[`CurrencySelector renders a disabled currency selector 1`] = `
<button
  class="tc__btn tc__currency-input__currency-selector-item tc__currency-input__currency-selector-item-disabled"
  data-selected="false"
  disabled=""
  type="button"
>
  <span
    class="tc__currency-input__currency-selector-item-content"
  >
    <svg
      class="tc__logo tc__logo-token-ETH"
      fill="none"
      height="40"
      viewBox="0 0 40 40"
      width="40"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20 40C31.0457 40 40 31.0457 40 20C40 8.9543 31.0457 0 20 0C8.9543 0 0 8.9543 0 20C0 31.0457 8.9543 40 20 40Z"
        fill="#627EEA"
      />
      <path
        clip-rule="evenodd"
        d="M11.25 20.275L20.6225 5V16.0875L11.25 20.275ZM20.6225 27.4588V34.9938L11.25 22.0201L20.6225 27.4588Z"
        fill="white"
        fill-rule="evenodd"
      />
      <path
        d="M20.625 25.7163L29.9962 20.275L20.625 16.09V25.7163Z"
        fill="white"
        fill-opacity="0.2"
      />
      <path
        clip-rule="evenodd"
        d="M20.625 5V16.0875L29.9962 20.275L20.625 5ZM20.625 27.4573V34.991L30.0025 22.0173L20.625 27.4573ZM20.6225 25.7163L11.25 20.275L20.6225 16.09V25.7163Z"
        fill="white"
        fill-opacity="0.6"
        fill-rule="evenodd"
      />
    </svg>
    <div
      class="tc__currency-input__currency-selector-item-label"
      style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(122, 122, 122); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      ETH
    </div>
  </span>
  <svg
    class="tc__icon tc__icon-down-chevron"
    fill="none"
    height="16"
    viewBox="0 0 16 16"
    width="16"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      clip-path="url(#clip0_909_5439)"
    >
      <path
        d="M15.7387 5.36992L8.63082 11.7636C8.54823 11.8385 8.44996 11.898 8.34169 11.9385C8.23342 11.9791 8.11729 12 8 12C7.88271 12 7.76658 11.9791 7.65831 11.9385C7.55004 11.898 7.45177 11.8385 7.36918 11.7636L0.261296 5.36992C0.0939908 5.21942 -2.49304e-09 5.01531 0 4.80248C2.49304e-09 4.58965 0.0939908 4.38554 0.261296 4.23504C0.428601 4.08455 0.655515 4 0.89212 4C1.12873 4 1.35564 4.08455 1.52294 4.23504L8 10.0693L14.4771 4.23504C14.6444 4.08455 14.8713 4 15.1079 4C15.3445 4 15.5714 4.08455 15.7387 4.23504C15.906 4.38554 16 4.58965 16 4.80248C16 5.01531 15.906 5.21942 15.7387 5.36992Z"
        fill="#062330"
      />
    </g>
    <defs>
      <clippath
        id="clip0_909_5439"
      >
        <rect
          fill="white"
          height="16"
          width="16"
        />
      </clippath>
    </defs>
  </svg>
</button>
`;
